src/java/org/apache/cassandra/tcm/transformations/AlterSchema.java

```diff
@@ -249,22 +253,56 @@ private static Map<ReplicationParams, Set<KeyspaceMetadata>> groupByReplication(
-        {
-            ImmutableMap<TableId, TableMigrationState> newTableStates =
-                prev.tableStates.entrySet().stream().filter(e -> !tableIds.contains(e.getKey())).collect(toImmutableMap());
-            return next.with(newTableStates);
-        }
-        else
+        MigrationState migrationState = prev;
+
+        Set<TableId> droppedIds = Streams.concat(altered.stream().flatMap(diff -> diff.tables.dropped.stream().map(TableMetadata::id)),
+                                                 dropped.stream().flatMap(ks -> ks.tables.stream().map(TableMetadata::id)))
+                                         .collect(toImmutableSet());
+
+        if (!droppedIds.isEmpty())
+            migrationState = migrationState.withMigrationsCompletedFor(droppedIds);
+
+        Set<TableId> completedIds = altered.stream().flatMap(diff -> diff.tables.altered.stream())
+                                           .filter(alt -> alt.before.params.transactionalMigrationFrom != TransactionalMigrationFromMode.none
+                                                          && alt.after.params.transactionalMigrationFrom == TransactionalMigrationFromMode.none)
+                                           .map(alt -> alt.after.id)
+                                           .collect(toImmutableSet());
+
+        if (!completedIds.isEmpty())
+            migrationState = migrationState.withMigrationsCompletedFor(completedIds);
+
+        // we treat explicitly switched migration types as a new migration here
+        Set<TableId> startedIds = altered.stream().flatMap(diff -> diff.tables.altered.stream())
+                                           .filter(alt -> alt.after.params.transactionalMigrationFrom != TransactionalMigrationFromMode.none
+                                                          && alt.before.params.transactionalMigrationFrom == alt.after.params.transactionalMigrationFrom)
+                                           .map(alt -> alt.after.id)
+                                           .collect(toImmutableSet());
+
+        if (!startedIds.isEmpty())
         {
-            return next;
+            List<Range<Token>> ranges;
+            Config.TransactionalRangeMigration migration = DatabaseDescriptor.getTransactionalRangeMigration();
+            switch (migration)
+            {
+                default: throw new IllegalStateException("Unhandled transactional range migration: " + migration);
+                case auto:
+                    Token minToken = DatabaseDescriptor.getPartitioner().getMinimumToken();
+                    ranges = Range.normalize(Collections.singletonList(new Range<>(minToken, minToken)));
+                    break;
+                case explicit:
+                    ranges = Collections.emptyList();
+                    break;
+            }
+
+            migrationState = migrationState.withRangesMigrating(startedIds, ranges, true);
```

Comment:
Also this is a correctness issue because if you don't precisely flip the ranges then you are basically migrating ranges to the new "from" that never actually finished migration.

Reply:
Actually I think this is safe/fine since all keys already migrated will be key migrated before being used. In fact you could reverse several times and it should be fine? You will migrate on one system before using the other and you are always going to see the latest value for a key because either you picked the correct system, or you did the key migration first.

Reply:
Although one random thought is that it's kind of important that the timestamps published by Paxos and Accord have the necessary barriers. IOW Accord should never write a value with a timestamp lower than what Paxos has used.